src/binary/cursor.rs

```diff
@@ -0,0 +1,1010 @@
+        uint::UInt,
+        var_int::VarInt,
+        var_uint::VarUInt,
+        IonTypeCode,
+    },
+    data_source::IonDataSource,
+    result::{decoding_error, illegal_operation, illegal_operation_raw, IonResult},
+    types::{IonType, SymbolId},
+};
+
+/// Information about the value over which the Cursor is currently positioned.
+#[derive(Clone, Debug)]
+struct CursorValue {
+    ion_type: IonType,
+    header: Header,
+    is_null: bool,
+    index_at_depth: usize,
+    length_in_bytes: usize,
+    last_byte: usize,
+    field_id: Option<SymbolId>,
+    annotations: Vec<SymbolId>,
+    parent_index: Option<usize>,
+}
+
+impl Default for CursorValue {
+    fn default() -> CursorValue {
+        CursorValue {
+            ion_type: IonType::Null,
+            header: Header {
+                ion_type: None,
+                ion_type_code: IonTypeCode::NullOrWhitespace,
+                length_code: length_codes::NULL,
+            },
+            field_id: None,
+            annotations: Vec::new(),
+            is_null: true,
+            index_at_depth: 0,
+            length_in_bytes: 0,
+            last_byte: 0,
+            parent_index: None,
+        }
+    }
+}
+
+// A low-level reader that offers no validation or symbol management.
+// It can only move and return the current value.
+pub struct BinaryIonCursor<R>
+where
+    R: IonDataSource,
+{
+    // The file, socket, array, or other BufRead implementor containing binary Ion bytes
+    data_source: R,
+    // Used for individual data_source.read() calls independent of input buffering
+    buffer: Vec<u8>,
+    // Tracks our position in the stream and information about the current value
+    cursor: CursorState,
+    // A jump table of pre-parsed header bytes
+    header_cache: Vec<IonResult<Option<Header>>>,
+}
+
+/* CursorState is broken out from the BinaryIonCursor struct to allow it to be cloned
+ * or replaced as part of a seek operation.
+ * See: https://github.com/amzn/ion-rust/issues/21
+ */
+#[derive(Clone, Debug)]
+pub struct CursorState {
+    // The (major, minor) version pair of the stream being read. Defaults to (1, 0)
+    ion_version: (u8, u8),
+    // How many bytes we've read from our data source
+    bytes_read: usize,
+    // How deeply nested the cursor is at the moment
+    depth: usize,
+    // The number of values that have been read at the current depth
+    index_at_depth: usize,
+    // Whether the cursor is currently traversing a struct's fields
+    is_in_struct: bool,
+    // Information about the value on which the cursor is currently sitting
+    value: CursorValue,
+    // All of the values into which the cursor has stepped. Empty at the top level.
+    parents: Vec<CursorValue>,
+}
+
+/// Verifies that the current value is of the expected type and that the bytes representing that
+/// value have not yet been consumed from the data source. This macro is called by the
+/// BinaryCursor#read_{typename} methods.
+macro_rules! read_safety_checks {
```

Comment:
Making this a macro allowed me to:
* Repeat the same `if` tests much more concisely throughout the code.
* Perform early returns from the call sites without requiring `if ... {return}` scaffolding.

Reply:
minor: naming of the macro seems off to me, perhaps, `read_type` or `read_val_of_type`